grails-app/controllers/uk/org/nbn/alerts/SavedSearchController.groovy

```diff
@@ -0,0 +1,236 @@
+package uk.org.nbn.alerts
+
+import au.ala.org.ws.security.RequireApiKey
+import au.org.ala.alerts.Notification
+import au.org.ala.alerts.User
+import grails.converters.JSON
+import org.apache.http.HttpStatus
+
+
+
+/**
+ * Controller for managing saved searches.
+ */
+class SavedSearchController {
+
+    static scaffold = SavedSearch
+
+    static allowedMethods = [
+        list: "GET",
+        save: "POST",
+        updateApi: "POST",
+        update: ["POST", "PUT"],
+        delete: "POST",
+        mySavedSearches: "GET"
+    ]
+
+    def savedSearchService
+    def authService
+
+    /**
+     * List all saved searches for the current user
+     *
+     * The method requires that the user is logged on, so that must be enforced
+     * eg by using annotation or however else the application enforces authentication
+     */
+    @RequireApiKey
+    def list() {
+        if (!params.userId) {
+            flash.errorMessage = "userId is a required parameter"
+            redirect(action: "index")
+            return
+        }
+
+        def user = User.findByUserId(params.userId)
+        if (!user) {
+            flash.errorMessage = "User not found"
+            redirect(action: "index")
+            return
+        }
+
+        def savedSearches = savedSearchService.getSavedSearch(params.userId)
+        render savedSearches as JSON
+    }
+
+    @RequireApiKey
+    def save() {
+        if (!params.userId) {
+            flash.errorMessage = "userId is a required parameter to save a search"
+            redirect(action: "mySavedSearches", params: [userId: params.userId])  // Add userId to redirect
+            return
+        }
+
+        def currentUser = User.findByUserId(params.userId)
+        if (!currentUser) {
+            flash.errorMessage = "User not found"
+            redirect(action: "mySavedSearches", params: [userId: params.userId])  // Add userId to redirect
+            return
+        }
+
+        try {
+            def savedSearch = savedSearchService.createSavedSearch(
+                params.userId,
+                params.name,
+                params.description,
+                params.searchRequestQueryUI
+            )
+
+            flash.message = "Search saved successfully"
+            redirect(action: "mySavedSearches", params: [userId: params.userId])  // This ensures we go back to My Saved Searches
+        } catch (Exception e) {
+            flash.errorMessage = "Error saving search: ${e.message}"
+            render(view: "create", model: [savedSearch: params])
+        }
+    }
+
+
+
+    @RequireApiKey
+    def updateApi() {
```

Comment:
only need one update method

can have 2 paths mapping to a single update method
can specify paths above the method itself using annotations
/api/update
/update

Reply:
deleted